### PR TITLE
Add qbs/1.18.0 recipe

### PR DESCRIPTION
**Pull request by rweickelt**
Specify library name and version:  **qbs/1.16.0**

- [x ] I've read the [guidelines](https://github.com/conan-io/conan-center-index/wiki#how-to-submit-a-pull-request) for contributing.
- [ x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [ x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [ x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

Qbs is a cross-platform build automation system.

This recipe extracts the Qbs binaries from an official Qt Creator binary
package provided by The Qt Company. It omits building Qbs from source
because that would require a reliably working Conan package for Qt.

Fixes: #1218

### CODE REVIEW

File: recipes/qbs/all/conanfile.py
Comment (madebr) [inline recipes/qbs/all/conanfile.py:10]:
```suggestion
    homepage = "http://qbs.io"
```

File: recipes/qbs/all/conanfile.py
Comment (madebr) [inline recipes/qbs/all/conanfile.py:14]:
```suggestion
        "arch": ["x86_64"],
        "os": ["Linux", "Macos", "Windows"]
```
The new cross building support of conan will map `os` to `os_build` if used as a build requirement.

File: recipes/qbs/all/conanfile.py
Comment (madebr) [inline recipes/qbs/all/conanfile.py:44]: The hooks will complain because `libexec` and `share` are not whitelisted.
Can these be moved?

Same goes for Macos.

File: recipes/qbs/all/conanfile.py
Comment (madebr) [inline recipes/qbs/all/conanfile.py:40]: Is it required to copy all qt5 libraries?
Do we want this?
Aren't these here for `qt creator`? (of which you're downloading the binaries in `build` but are not copying the binaries)
On Linux, running `objdump -p` on `qbs` returns these non-system libraries:
`libqbscore.so.1.16` and `libQt5Core.so.5`

File: recipes/qbs/all/conanfile.py
Comment (madebr) [inline recipes/qbs/all/conanfile.py:30]: These files belong to qt creator.
Are these needed?
Will qt creator use this folder when using this conan qbs package?

File: recipes/qbs/all/conandata.yml
Comment (madebr) [inline recipes/qbs/all/conandata.yml:5]: Can you use sha256 hashes?
md5 hashes are considered insecure nowadays.

Comment (rweickelt) [reply recipes/qbs/all/conandata.yml:5]: Sure. I thought this is only about integrity check and only the md5sums were published on the Qt download page.

File: recipes/qbs/all/conanfile.py
Comment (rweickelt) [reply recipes/qbs/all/conanfile.py]: Done.

File: recipes/qbs/all/conanfile.py
Comment (rweickelt) [reply recipes/qbs/all/conanfile.py]: Done.

File: recipes/qbs/all/conanfile.py
Comment (rweickelt) [reply recipes/qbs/all/conanfile.py]: I don't think they can be moved. The structure is defined at compile time.

**Comment by madebr** [comment]
It would be nice if qbs would be built from source, like the cmake recipe.

It looks like qbs can be built using qmake: https://github.com/qbs/qbs/tree/master/scripts
But qmake requires qt5, which is not available (yet) on cci, but is on bincrafters: https://github.com/bincrafters/conan-qt

File: recipes/qbs/all/conanfile.py
Comment (rweickelt) [reply recipes/qbs/all/conanfile.py]: > Is it required to copy all qt5 libraries?
> Do we want this?

No, it is not required to copy all qt5 libraries, but the ones listed above were necessary because this package has been compiled with config-ui enabled. That causes a dependency on libQt5Gui and a couple of others and of course the platform plugins. Maybe I have overlooked 2 or 3.

It is not ideal, but building Qbs from source in a Conan environment is not something I have time for.

Actually, we can just say that using `qbs config-ui` is forbidden and remove a lot of the libraries. I guess nobody would do that anyway in a Conan environment. I have updated the PR accordingly.

> Aren't these here for qt creator? (of which you're downloading the binaries in build but are not copying the binaries)

No, they are needed by `libqbscore` or one of its dependencies.

> On Linux, running objdump -p on qbs returns these non-system libraries:
libqbscore.so.1.16 and libQt5Core.so.5

Rather use `ldd` to find out run-time dependencies (including transitive ones). This does still not print the required plugins though.

File: recipes/qbs/all/conanfile.py
Comment (rweickelt) [reply recipes/qbs/all/conanfile.py]: > These files belong to qt creator. Are these needed?

The reason why you see a `qtcreator` in the path is that Qbs is compiled together with Qt Creator and has become a more or less integral part. We could probably patch the rpath of the binary and move all the libaries to `lib`, but I don't think we would gain anything from that.

**Comment by rweickelt** [comment]
> It would be nice if qbs would be built from source, like the cmake recipe.

Once Qt is available in conan-center we can talk.

> It looks like qbs can be built using qmake: https://github.com/qbs/qbs/tree/master/scripts

I would prefer to build it from source using Qbs because that gives us more control over the build process. That's why we need a seed package first.

> But qmake requires qt5, which is not available (yet) on cci, but is on bincrafters: https://github.com/bincrafters/conan-qt

That requires everybody to rebuild Qt from source because the required configuration (with QtScript) is not available on bintray.

I am generally open to building Qbs from source, but I'd like to have Qt in conan-center first and would like to have a Qbs package as a seed. That gives us much more flexibility than a qmake build.

**Comment by madebr** [comment]
> That requires everybody to rebuild Qt from source because the required configuration (with QtScript) is not available on bintray.
> 
> I am generally open to building Qbs from source, but I'd like to have Qt in conan-center first and would like to have a Qbs package as a seed. That gives us much more flexibility than a qmake build.

I am just a contributor here, not in control of anything.
So it's up to the reviewers to decide.
I will @ericLemanissier for his ideas.

**Comment by ericLemanissier** [comment]
I don't know qbs, but i'm surprised that it uses qtscript, which has not been maintained for years. Regarding the integration of qt in cci, it is blocked until a virtual package for x11 client libraries is created

**Comment by rweickelt** [comment]
> which has not been maintained for years

It is deprecated, but still a valid part of the Qt offering in Qt5. We are porting Qbs to QML, but that won't be ready before Qt6.

File: recipes/qbs/all/conandata.yml
Comment (Croydon) [inline recipes/qbs/all/conandata.yml:3]: The `conandata.yml` file does not allows custom fields, so none of this will work

Comment (rweickelt) [reply recipes/qbs/all/conandata.yml:3]: [But it works](https://travis-ci.com/github/rweickelt/conan-center-index).

[And here is why](https://docs.conan.io/en/latest/reference/config_files/conandata.yml.html):
> This file can be used, for example, to declare a list of sources links and checksums for the recipe or a list patches to apply to them, but you can use it to store any data you want to extract from the recipe. For example: 

Or am I missing something?

**Comment by rweickelt** [comment]
> It might be better to add such recipe to Bincrafters first, depend on the Qt recipe and build from source

Is it a hard requirement for to build packages from source?

**Comment by Croydon** [comment]
No, for installer it can be okay to repackage official binaries, but we should prevent it when possible

File: recipes/qbs/all/conandata.yml
Comment (rweickelt) [reply recipes/qbs/all/conandata.yml]: OK, it's a restriction by conan-center. Should be fine now.

File: recipes/qbs/all/conanfile.py
Comment (rweickelt) [reply recipes/qbs/all/conanfile.py]: OK, I found a FAQ entry which says that all content of binary tools should be placed in bin. Done.

File: recipes/qbs/all/conanfile.py
Comment (rweickelt) [reply recipes/qbs/all/conanfile.py]: Marking as resolved. The latest commit includes only the absolutely required libraries.

File: recipes/qbs/all/conanfile.py
Comment (rweickelt) [reply recipes/qbs/all/conanfile.py]: Marking as resolved since I think the question was answered. The latest commits put all files under bin.

**Comment by rweickelt** [comment]
I have [checked this recipe on all 3 major platforms with the conan-center hook installed](https://travis-ci.com/github/rweickelt/conan-center-index/builds/168475446
) and resolved all remaining errors. Therefore, I am confident that it will work in the CI as well.

What is required to take this PR forward? What is the next step? Are there any other formal requirements that need to be fulfilled? Maybe @uilianries can help? Thanks.

File: recipes/qbs/all/conanfile.py
Comment (uilianries) [inline recipes/qbs/all/conanfile.py:2]:
```suggestion
from conans.errors import ConanInvalidConfiguration
import os
```

File: recipes/qbs/all/conanfile.py
Comment (uilianries) [inline recipes/qbs/all/conanfile.py:18]:
```suggestion
    settings = "arch", "os"
```

It's better using configure + ConanInvalidConfiguration. Here, it will result in a ConanException and will break the CI job.

File: recipes/qbs/all/conanfile.py
Comment (uilianries) [inline recipes/qbs/all/conanfile.py:23]:
```suggestion

    def configure(self):
        if self.settings.os not in ["Windows", "Linux", "Macos"]:
        	raise ConanInvalidConfiguration("The OS ({}) is not supported by {}.".format(self.settings.os, self.name))
        if self.settings.arch not in ["x86_64"]:
          	raise ConanInvalidConfiguration("The arch ({}) is not supported by {}.".format(self.settings.arch, self.name))

```

File: recipes/qbs/all/conanfile.py
Comment (uilianries) [inline recipes/qbs/all/conanfile.py:54]: share folder is not allowed in CCI. Remove it.

File: recipes/qbs/all/conanfile.py
Comment (uilianries) [inline recipes/qbs/all/conanfile.py:33]:
```suggestion
        binpath = os.path.join(self.package_folder, self._data['bin_folder'])
        self.output.info("Appending PATH env var with : " + binpath)
```

File: recipes/qbs/all/test_package/conanfile.py
Comment (uilianries) [inline recipes/qbs/all/test_package/conanfile.py:10]:
```suggestion
    settings = "os", "arch"
```

File: recipes/qbs/all/conanfile.py
Comment (uilianries) [inline recipes/qbs/all/conanfile.py:26]: Add `7zip/19.00@` as build requirement. 7z is not a default app.

File: recipes/qbs/all/conanfile.py
Comment (uilianries) [inline recipes/qbs/all/conanfile.py:34]:
```suggestion
        self.env_info.PATH.append(binpath)
```

**Comment by rweickelt** [comment]
> I'm not confident with this package, I would prefer having it on Bincrafters instead.

Thanks for the review. It was my understanding that CCI is the successor of the Bincrafters project and the place where all packages should end up nowadays. Also, I would like to leverage the common infrastructure on CCI instead of maintaining my own one.

> It requires Qt Creator, so you have 2 "packages" in the same one. It's not correct and it's not a good approach.

As the PR description says: It extracts Qbs from a Qt Creator installer. No content of Qt Creator ends up in the resulting Qbs Conan package.
 
> Bincrafters provides the modular Qt (thanks to @ericLemanissier), which can be used. On CCI we would need Qt first ...

It looks like @ericLemanissier is working very hard to get the Qt package on CCI. So there is a good chance that I can build Qbs from source on CCI. @ericLemanissier, what is your plan with #1759? Are you going to make this production-ready during the next few weeks?

File: recipes/qbs/all/conanfile.py
Comment (rweickelt) [reply recipes/qbs/all/conanfile.py]: Done.

File: recipes/qbs/all/conanfile.py
Comment (rweickelt) [reply recipes/qbs/all/conanfile.py]: Done

File: recipes/qbs/all/conanfile.py
Comment (rweickelt) [reply recipes/qbs/all/conanfile.py:23]: Done

File: recipes/qbs/all/conanfile.py
Comment (rweickelt) [reply recipes/qbs/all/conanfile.py:54]: This is already copied to the bin folder as suggested. `'dst': 'bin'`

File: recipes/qbs/all/conanfile.py
Comment (rweickelt) [reply recipes/qbs/all/conanfile.py]: Done

File: recipes/qbs/all/test_package/conanfile.py
Comment (rweickelt) [reply recipes/qbs/all/test_package/conanfile.py:8]: Done

File: recipes/qbs/all/conanfile.py
Comment (rweickelt) [reply recipes/qbs/all/conanfile.py:26]: This uses the 7z module that is baked into cmake which works on all platforms. `Add 7zip/19.00@` is only available for Windows.

File: recipes/qbs/all/conanfile.py
Comment (rweickelt) [reply recipes/qbs/all/conanfile.py]: Done.

**Comment by ericLemanissier** [comment]
I hava no idea when the recipe will be ready for CCI. There are probably a lot of obstacles along the way.

**Comment by SSE4** [comment]
```
qbs: error while loading shared libraries: libGL.so.1: cannot open shared object file: No such file or directory
```
this is weird. build system requires OpenGL?
@rweickelt could you double-check if it's a real dependency or just bug?
if it's real, you may try to reference `opengl/system` recipe as a requirement.

**Comment by rweickelt** [comment]
> ```
> qbs: error while loading shared libraries: libGL.so.1: cannot open shared object file: No such file or directory
> ```
> 
> this is weird. build system requires OpenGL?
> @rweickelt could you double-check if it's a real dependency or just bug?
> if it's real, you may try to reference `opengl/system` recipe as a requirement.

Thanks, @SSE4. That is because this recipe does not build anything, but just repackage an existing installer. I've decided to no longer follow that route but to wait until #1759 is merged.

**Comment by rweickelt** [comment]
@uilianries, I am a bit lost with the error message. Does the conan-center CI not build dependencies? Also, can I disable Debug builds since this is a binary tool?

File: recipes/qbs/all/conanfile.py
Comment (ericLemanissier) [inline recipes/qbs/all/conanfile.py:37]: this cannot be done. if certain qt options absolutely need to have a given value, then what should be done instead is to check the value in `validate(self)` and raise `ConanInvalidConfiguration` if the value is incorrect

**Review by ericLemanissier — CHANGES_REQUESTED** [review]

File: recipes/qbs/all/conanfile.py
Comment (rweickelt) [reply recipes/qbs/all/conanfile.py:37]: @ericLemanissier, pardon, I don't understand, could you elaborate? This recipe requires a static Qt build with QtScript. Since it is static, I'd like to reduce Qt even further, but that is not strictly needed.

I tried this:
```
    def validate(self):
        if self.options["qt"].gui != False:
            raise ConanInvalidConfiguration("Option gui not supported")
        if self.options["qt"].qtscript != True:
            raise ConanInvalidConfiguration("Option qtscript is required")
        if self.options["qt"].widgets != False:
            raise ConanInvalidConfiguration("Option widgets not supported")
```
and got:
```
ERROR: qbs/1.18.0: Error in validate() method, line 54
	if self.options["qt"].gui != False:
	ConanException: option 'gui' doesn't exist
Possible options are []
```

Comment (ericLemanissier) [reply recipes/qbs/all/conanfile.py:37]: this could be because you are using qt as a build_requires whereas it is a requirement

Comment (rweickelt) [reply recipes/qbs/all/conanfile.py:37]: @ericLemanissier, now I am even more confused. Qt is only a build requirement for Qbs. Since this recipe does static linking, Qbs does not have any run-time dependencies on Qt.

Comment (ericLemanissier) [reply recipes/qbs/all/conanfile.py:37]: the fact that it links statically does not change the fact that qt is used at run-time by qbs. on the other hand, you probably want to use qbs as a build requirement.
Does qbs work if you remove all the checks on qt options ?

File: recipes/qbs/all/conanfile.py
Comment (mabean) [inline recipes/qbs/all/conanfile.py:103]: @rweickelt can you also remove arch option or better use arch_build? Same for os option. You'll get troubles with cross-building profiles like android.

```
[settings]
arch=armv7
arch_build=x86_64
build_type=Release
compiler=clang
compiler.libcxx=libc++
compiler.version=9
os=Android
os.api_level=21
os_build=Windows
[options]
[build_requires]
*: android_ndk_installer/r21b
[env]
```

File: recipes/qbs/all/conanfile.py
Comment (rweickelt) [reply recipes/qbs/all/conanfile.py:37]: > Does qbs work if you remove all the checks on qt options ?

It does not, because the prebuilt Qt packages on conan-center do not come with QtScript enabled. These are the minimum requirements I have:

        self.options["qt"].gui = False
        self.options["qt"].qtscript = True
        self.options["qt"].widgets = False

How would I implement that in the Qbs recipe?

> the fact that it links statically does not change the fact that qt is used at run-time by qbs.

Yes it is needed at run-time, but let me use different wording: There is no run-time dependency of Qbs on files from the Qt package. Qt is baked into the Qbs application binaries and thus, the Qbs package becomes 100% self-containing. Even if I linked Qt dynamically I would include the Qt libs into the Qbs package. The Qt package configuration used by Qbs must never ever leak to consumers of the Qbs package.

>  on the other hand, you probably want to use qbs as a build requirement.

Not probably. You always want to use Qbs as a build requirement. Would you ever want CMake as a (non-build) requirement in your project?

File: recipes/qbs/all/conanfile.py
Comment (rweickelt) [reply recipes/qbs/all/conanfile.py:103]: Have a look at this comment: https://github.com/conan-io/conan-center-index/pull/1716#discussion_r429558736

Comment (mabean) [reply recipes/qbs/all/conanfile.py:103]: Missed this comment, sorry

File: recipes/qbs/all/conanfile.py
Comment (prince-chrismc) [inline recipes/qbs/all/conanfile.py:51]: This list seems way above the C++11 requirement

File: recipes/qbs/all/conanfile.py
Comment (ericLemanissier) [reply recipes/qbs/all/conanfile.py:37]: unfortunately, CCI only builds the recipes with default options (the only variation being the `shared` option), so the only solution to get qbs in CCI would be to set the default value of qt:qtscript to True in the qt recipe first.
By the way, this dependency kind of makes QBS dead on arrival, because qtscript does not exist any more in qt6 :/

File: recipes/qbs/all/conanfile.py
Comment (rweickelt) [reply recipes/qbs/all/conanfile.py:51]: What would be your suggestion? We are bound by https://doc.qt.io/qt-5/supported-platforms.html

File: recipes/qbs/all/conanfile.py
Comment (rweickelt) [reply recipes/qbs/all/conanfile.py:37]: > unfortunately, CCI only builds the recipes with default options (the only variation being the shared option), so the only solution to get qbs in CCI would be to set the default value of qt:qtscript to True in the qt recipe first.

that was the missing piece.

> By the way, this dependency kind of makes QBS dead on arrival, because qtscript does not exist any more in qt6 :/

You don't need to build Qbs with Qt6 in order to build Qt6 projects with it. In the meantime we are heavily working on a QJS port. Unfortunately the Qt company has abandoned QtScript without providing a sufficient replacement API, so porting is not trivial.

Comment (rweickelt) [reply recipes/qbs/all/conanfile.py:37]: > so the only solution to get qbs in CCI would be to set the default value of qt:qtscript to True in the qt recipe first.

@ericLemanissier, great, would that be acceptable? But maybe tomorrow somebody else needs qtdeclarative emabled. On the other hand, how would one ever bring a project to cci that needs Qt modules beyond qtbase?

Comment (ericLemanissier) [reply recipes/qbs/all/conanfile.py:37]: I have no idea if it's acceptable, or if CI time/disk space is sufficient to build qtbase+qtscript. You'd have to make a PR to know.

File: recipes/qbs/all/conanfile.py
Comment (ericLemanissier) [inline recipes/qbs/all/conanfile.py:20]:
```suggestion
    def build_requirements(self):
        self.build_requires("qt/5.15.2")
        if tools.os_info.is_windows and self.settings.compiler == "Visual Studio":
```

**Comment by ericLemanissier** [comment]
@SSE4 the `blocked` label can be removed, now that qt is on CCI

File: recipes/qbs/all/conanfile.py
Comment (prince-chrismc) [reply recipes/qbs/all/conanfile.py:20]: bump https://github.com/conan-io/conan-center-index/pull/1716/files#r592978685

**Pull request closed without merging**
